### ISSUE TITLE
Why was zap.Logger interface replaced with a struct?

### ISSUE THREAD

**Issue by yurishkuro** [uber-go/zap#381]
Q: Earlier versions defined `zap.Logger` as an interface. Now it seems to have been replaced with a struct, so we'd need to change all of the code to accept `*zap.Logger` or define our own interface. The latter isn't too bad, but it means we would either need to redirect all functions like `zap.String()` or use two different namespaces.

**Comment by akshayjshah** [comment]
Great question - I have a feeling I'll be referring to this issue in the future 😉 

I changed to a concrete type for a few reasons:
1. `zapcore.Core` is the fundamental interface; if you want to replace the heart of zap's implementation with something different, that's your extension point. The `zap.WrapCore` option makes this easy.
2. Concrete types are much easier to extend in minor releases - I can't add to an interface without cutting a new major version.
3. A `zap.Logger` interface isn't terribly useful, since it's a pretty large API that frequently references other zap types. Good interfaces usually deal *mostly* in types from the standard library.

I don't follow your point about separate namespaces - can you elaborate? You're free to do something like this:

```
type Logger interface {
  Debug(string, ...zapcore.Field)
  Info(string, ...zapcore.Field)
}

func doStuff(log Logger) {
  log.Info("running", zap.String("foo", "bar"))
  // your code here
}

func main() {
  doStuff(zap.NewNop())
  doStuff(newAlternateLoggerImplementation())
}
```

This should work out of the box. Obviously, it doesn't actually let you sever your dependency on zap, since fields are values you'll have to construct somehow.

I don't think it's onerous to migrate to accepting a pointer - `gofmt -w -r "zap.Logger->*zap.Logger" .` will auto-migrate a whole project.

**Comment by yurishkuro** [comment]
> I don't follow your point about separate namespaces - can you elaborate?

What I meant was, say we define and use our own interface `jaeger-lib/zap/Logger` and update all imports. It won't work because the same code is using zap.Fields. We can either add field constructors to `jaeger-lib/zap` that will redirect to real constructors, or we'd have to pick a different namespace for the interface, e.g. `jaeger-lib/jzap/Logger` (which would require larger code changes, not sure if `gofmt -r` would handle it, I never tried).

> I don't think it's onerous to migrate to accepting a pointer

That goes against "accept interfaces return structs" principle.

I think our best bet is to try to define our own narrower interface. I'll update this with the results.

**Comment by akshayjshah** [comment]
> or we'd have to pick a different namespace for the interface, e.g. jaeger-lib/jzap/Logger (which would require larger code changes, not sure if gofmt -r would handle it, I never tried).

It will :) This is the approach I used to start migrating Jaeger's backend; most of the migration was doable with a handful of sed and gofmt commands. I stopped when I realized that updating also requires maintaining the interop layer with the zap variant in opentracing-go.

> That goes against "accept interfaces return structs" principle.

It does. I generally like that advice, but don't think it's useful when the type in question references many other types in the same package - as you've noted, treating it as an interface requires copying and shimming large portions of the package. If that's a good tradeoff for the Jaeger codebase, though, go for it!
